rls: only reset backoff on recovery from TRANSIENT_FAILURE

ulascansenturk opened the pull request:
Fix control channel connectivity monitoring to track TRANSIENT_FAILURE state explicitly. Only reset backoff timers when transitioning from TRANSIENT_FAILURE to READY, not for benign state changes like READY → IDLE → READY.

RELEASE NOTES: N/A

Fixes  #8693

ulascansenturk committed:
rls: only reset backoff on recovery from TRANSIENT_FAILURE

Fix control channel connectivity monitoring to track TRANSIENT_FAILURE
state explicitly. Only reset backoff timers when transitioning from
TRANSIENT_FAILURE to READY, not for benign state changes like
READY → IDLE → READY.

Fixes #8693

eshitachandwani commented on balancer/rls/control_channel_test.go:
```diff
@@ -463,3 +465,94 @@ func (s) TestNewControlChannelUnsupportedCredsBundle(t *testing.T) {
+			}
+
+			// Give extra time for any pending callbacks
+			time.Sleep(100 * time.Millisecond)
```
I dont think adding `time.Sleep()` to wait for state changes is a good idea, since it can make tests flake if sometimes the state transitions take longer, we should look for better guarantees to make sure the states have transitioned. 
cc : @easwars

ulascansenturk committed:
address reviews

ulascansenturk committed:
address reviews

ulascansenturk committed:
Shorter comment

ulascansenturk committed:
Fix test synchronization and remove time.Sleep dependencies

- Add testOnlyInitialReadyDone channel for proper test synchronization
- Signal when monitoring goroutine processes initial READY state
- Tests wait for this signal instead of using time.Sleep
- All synchronization now uses channels/callbacks - no arbitrary sleeps
- Tests pass consistently with race detector

Addresses review feedback about removing time.Sleep for state transitions.

easwars commented on balancer/rls/control_channel.go:
FYI: This is the exact text that describes the expected behavior:

The policy will monitor the state of the control plane channel.  When the state transitions to TRANSIENT_FAILURE, it will record that transition, and the next time it transitions to READY, the policy will iterate through the cache to reset the backoff timeouts in all cache entries.  Specifically, this means that it will reset the backoff state and cancel the pending backoff timer.  Note that when cancelling the backoff timer, just like when the backoff timer fires normally, a new picker is returned to the channel, to force it to re-process any wait-for-ready RPCs that may still be queued if we failed them while we were in backoff.  However, we should optimize this case by returning only one new picker, regardless of how many backoff timers are cancelled.

ulascansenturk committed:
Fix comments

ulascansenturk committed:
address reviews

ulascansenturk committed:
gofmt

easwars commented on balancer/rls/control_channel.go:
```diff
@@ -57,24 +57,22 @@ type controlChannel struct {
+	logger               *internalgrpclog.PrefixLogger
+	unsubscribe          func()
+	seenTransientFailure bool
+	mu                   sync.Mutex
```
Nit: Please group the mutex and the fields that it guards as a separate block. It is a generally used convention that any fields that come right below a mutex are to be guarded by that mutex.

easwars commented on balancer/rls/control_channel_test.go:
```diff
@@ -463,3 +465,119 @@ func (s) TestNewControlChannelUnsupportedCredsBundle(t *testing.T) {
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
To be able to test this in an e2e style, we would need to do make it possible for the test to see the connectivity state changes on the control channel, but without adding hooks into the test code. I propose the following:

- Add a new variable that can be overridden by the test. This variable will point to a function that returns the subscriber to the passed to the `SubscribeToConnectivityStateChanges` API
```golang
var newConnectivityStateSubscriber = connStateSubscriber
```

- Set the above variable to the following piece of code in `control_channel.go`
```golang
func connStateSubscriber(sub grpcsync.Subscriber) grpcsync.Subscriber {
       return sub
}
```

- When calling the `SubscribeToConnectivityStateChanges` API, use the above function as follows:
```golang
       ctrlCh.unsubscribe = internal.SubscribeToConnectivityStateChanges.(func(cc *grpc.ClientConn, s grpcsync.Subscriber) func())(ctrlCh.cc, newConnectivityStateSubscriber(ctrlCh))
```

- In the test, create an implementation of the `grpcsync.Subscriber` interface that contains a delegate. This makes it possible for the test to delegate to the subscriber set by the non-test code, but make the connectivity state changes available to the test:
```golang
type wrappingConnectivityStateSubscriber struct {
       delegate    grpcsync.Subscriber
       connStateCh chan connectivity.State
}

func (w *wrappingConnectivityStateSubscriber) OnMessage(msg any) {
       w.delegate.OnMessage(msg)
       w.connStateCh <- msg.(connectivity.State)
}
```

- Modify `TestControlChannelConnectivityStateMonitoring` or add a new test where we do the override. You would probably want to use a `testutils.
```golang
       // Override the connectivity state subscriber.
       wrappedSubscriber := &wrappingConnectivityStateSubscriber{connStateCh: make(chan connectivity.State, 1)}
       origConnectivityStateSubscriber := newConnectivityStateSubscriber
       newConnectivityStateSubscriber = func(delegate grpcsync.Subscriber) grpcsync.Subscriber {
               wrappedSubscriber.delegate = delegate
               return wrappedSubscriber
       }
       defer func() { newConnectivityStateSubscriber = origConnectivityStateSubscriber }()
```

- In the test body, we need to check the appropriate state transitions in a bunch of places:
```golang
	// Make sure an RLS request is sent out.
	verifyRLSRequest(t, rlsReqCh, true)

	// Verify that the control channel moves to READY.
	wantStates := []connectivity.State{
		connectivity.Connecting,
		connectivity.Ready,
	}
	for _, wantState := range wantStates {
		select {
		case gotState := <-wrappedSubscriber.connStateCh:
			if gotState != wantState {
				t.Fatalf("Unexpected connectivity state: got %v, want %v", gotState, wantState)
			}
		case <-ctx.Done():
			t.Fatalf("Timeout waiting for RLS control channel to become %q", wantState)
		}
	}

	// Stop the RLS server.
	lis.Stop()

	// Verify that the control channel moves to IDLE.
	wantStates = []connectivity.State{
		connectivity.Idle,
	}
	for _, wantState := range wantStates {
		select {
		case gotState := <-wrappedSubscriber.connStateCh:
			if gotState != wantState {
				t.Fatalf("Unexpected connectivity state: got %v, want %v", gotState, wantState)
			}
		case <-ctx.Done():
			t.Fatalf("Timeout waiting for RLS control channel to become %q", wantState)
		}
	}

	// Make another RPC similar to the first one. Since the above cache entry
	// would have expired by now, this should trigger another RLS request. And
	// since the RLS server is down, RLS request will fail and the cache entry
	// will enter backoff, and we have overridden the default backoff strategy to
	// return a value which will keep this entry in backoff for the whole duration
	// of the test.
	makeTestRPCAndVerifyError(ctx, t, cc, codes.Unavailable, nil)

	// Verify that the control channel moves to TRANSIENT_FAILURE.
	wantStates = []connectivity.State{
		connectivity.Connecting,
		connectivity.TransientFailure,
	}
	for _, wantState := range wantStates {
		select {
		case gotState := <-wrappedSubscriber.connStateCh:
			if gotState != wantState {
				t.Fatalf("Unexpected connectivity state: got %v, want %v", gotState, wantState)
			}
		case <-ctx.Done():
			t.Fatalf("Timeout waiting for RLS control channel to become %q", wantState)
		}
	}

	// Restart the RLS server.
	lis.Restart()
```

The above will test the READY --> TF --> READY transition.

For the READY --> IDLE --> READY, we need to restart the RLS server once the control channel goes IDLE, and then wait for it to go READY before attempting another RPC and verifying that backoffs are not reset.

Let me know what you think about this approach.

Thanks

ulascansenturk committed:
address reviews

ulascansenturk committed:
fix

ulascansenturk committed:
addres easwars's comment

ulascansenturk committed:
fixes

ulascansenturk committed:
Merge remote-tracking branch 'origin' into fix/8693-rls-control-channel-state-monitoring

ulascansenturk committed:
address comments

easwars commented on balancer/rls/balancer_test.go:
```diff
@@ -1016,6 +1095,157 @@ func (s) TestControlChannelConnectivityStateMonitoring(t *testing.T) {
+		select {
+		case gotState := <-wrappedSubscriber.connStateCh:
+			if gotState == connectivity.Ready {
+				goto initialReadyReached
```
Please get rid of gotos in this test as well.

easwars commented on balancer/rls/balancer_test.go:
```diff
@@ -1016,6 +1095,157 @@ func (s) TestControlChannelConnectivityStateMonitoring(t *testing.T) {
+	select {
+	case <-resetBackoffCalled:
+		t.Fatal("Backoff reset was triggered for initial READY state, want no reset")
+	default:
```
Please use a short timeout here instead of the default case.

easwars commented on balancer/rls/balancer_test.go:
```diff
@@ -1016,6 +1095,157 @@ func (s) TestControlChannelConnectivityStateMonitoring(t *testing.T) {
+				// This test is specifically for IDLE transitions. If we hit
+				// TRANSIENT_FAILURE, the other test covers that case.
+				t.Skip("Control channel went to TRANSIENT_FAILURE instead of IDLE; skipping IDLE transition test")
+			}
```
I don't think this is possible. When a connected subchannel breaks (i.e., when we stop the listener above, the grpc client will notice the connection to the backend is broken), it moves to IDLE and only starts connecting when the subchannel is asked to connect. 

In this test, we are not trying another RPC after stopping the server. So, there will not be an attempt to connect on the broken subchannel.

Please correct me if you think otherwise.

easwars commented on balancer/rls/balancer_test.go:
```diff
@@ -1016,6 +1095,157 @@ func (s) TestControlChannelConnectivityStateMonitoring(t *testing.T) {
+		select {
+		case gotState := <-wrappedSubscriber.connStateCh:
+			if gotState == connectivity.Ready {
+				goto readyAfterIdle
```
And here.

ulascansenturk committed:
address review comments: replace goto with labeled for+break, use short timeouts

- Replace goto/label patterns with labeled for loops and break
- Replace default cases with time.After(10ms) for proper timing
- Remove impossible TRANSIENT_FAILURE handling in IDLE test

ulascansenturk committed:
extract waitForConnectivityState helper to reduce duplication

Replace 5 identical labeled-for-select loops with a shared helper
function that waits for a specific connectivity state on the channel.
